pkg/instancetype/controller/vm/controller_test.go

```diff
@@ -0,0 +1,1052 @@
+			vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+				Name: instancetypeObj.Name,
+				Kind: instancetypeapi.SingularResourceName,
+			}
+
+			var err error
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+			Expect(err).To(Succeed())
+
+			expectedRevisionName := instancetype.GetRevisionName(
+				vm.Name, instancetypeObj.Name, instancetypeObj.GroupVersionKind().Version, instancetypeObj.UID, instancetypeObj.Generation)
+			expectedRevision, err := instancetype.CreateControllerRevision(vm, instancetypeObj)
+			Expect(err).ToNot(HaveOccurred())
+
+			sanitySync(vm, vmi)
+
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+			Expect(err).To(Succeed())
+			Expect(vm.Spec.Instancetype.RevisionName).To(Equal(expectedRevision.Name))
+
+			revision, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(
+				context.Background(), expectedRevisionName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			revisionInstancetype, ok := revision.Data.Object.(*v1beta1.VirtualMachineInstancetype)
+			Expect(ok).To(BeTrue(), "Expected Instancetype in ControllerRevision")
+
+			Expect(revisionInstancetype.Spec).To(Equal(instancetypeObj.Spec))
+		})
+
+		DescribeTable("should apply VirtualMachineInstancetype from ControllerRevision to VirtualMachineInstance",
+			func(getRevisionData func() []byte) {
+				instancetypeRevision := &appsv1.ControllerRevision{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "crName",
+					},
+					Data: runtime.RawExtension{
+						Raw: getRevisionData(),
+					},
+				}
+
+				instancetypeRevision, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(
+					context.Background(), instancetypeRevision, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+					Name:         instancetypeObj.Name,
+					Kind:         instancetypeapi.SingularResourceName,
+					RevisionName: instancetypeRevision.Name,
+				}
+				vm, err = virtClient.VirtualMachine(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+				Expect(err).To(Succeed())
+
+				Expect(instancetypeController.ApplyToVMI(vm, vmi)).To(Succeed())
+
+				Expect(vmi.Spec.Domain.CPU.Sockets).To(Equal(instancetypeObj.Spec.CPU.Guest))
+				Expect(*vmi.Spec.Domain.Memory.Guest).To(Equal(instancetypeObj.Spec.Memory.Guest))
+				Expect(vmi.Annotations).To(HaveKeyWithValue(virtv1.InstancetypeAnnotation, instancetypeObj.Name))
+				Expect(vmi.Annotations).ToNot(HaveKey(virtv1.PreferenceAnnotation))
+				Expect(vmi.Annotations).ToNot(HaveKey(virtv1.ClusterInstancetypeAnnotation))
+				Expect(vmi.Annotations).ToNot(HaveKey(virtv1.ClusterPreferenceAnnotation))
+			},
+			Entry("using v1alpha1 and VirtualMachineInstancetypeSpecRevision with APIVersion", func() []byte {
+				v1alpha1instancetypeSpec := v1alpha1.VirtualMachineInstancetypeSpec{
```

Comment:
Would be possible to extract the v1alpha1.VirtualMachineInstancetypeSpec struct to somewhere and then just reuse it in other tests?

Reply:
I'd rather keep the small amount of duplication in the test entries to make to clear to the reader what's being used here tbh.